### PR TITLE
Consider changing assert! to debug_assert! when it calls visit_with

### PR DESCRIPTION
**Pull request by ljedrz**
The perf run from #52956 revealed that there were 3 benchmarks that benefited most from changing `assert!`s to `debug_assert!`s:

- issue #46449: avg -4.7% for -check
- deeply-nested (AKA #38528): avg -3.4% for -check
- regression #31157: avg -3.2% for -check

I analyzed their fixing PRs and decided to look for potentially heavy assertions in the files they modified. I noticed that all of the non-trivial ones contained indirect calls to `visit_with()`.

It might be a good idea to consider changing `assert!` to `debug_assert!` in those places in order to get the performance wins shown by the benchmarks.

### CODE REVIEW

**Comment by varkor** [comment]
Although the perf improvements are nice, I'd want to be sure that this assertions have not been hit in practice (or would definitely be caught by another, informative assertion), because this assertions can often be very useful.

I'm not quite sure we'd establish this, other than reasoning about the code. Perhaps it'd be possible instead to pull these assertions further out of the loops they're in (possibly keeping the `debug_assert!`s where they are, but adding additional `asserts` that would trigger if something went wrong, potentially at a slightly less informative location).

**Comment by varkor** [comment]
Okay, as these conditions are non-trivial with regards to performance and it seems likely to me that if they fail, we're going to hit other assertions (later on, but assuming we can reproduce the issue, that should be sufficient), I think it's probably fine to change these. Thanks for looking into it!

@bors r+ rollup

**Comment by kennytm** [comment]
@bors r-

Merge conflict in `src/librustc/traits/select.rs`.

**Comment by ljedrz** [comment]
Rebased.

File: src/librustc/ty/layout.rs [hunk 1]
```diff
@@ -1300,7 +1300,7 @@ impl<'a, 'tcx> SizeSkeleton<'tcx> {
                 let tail = tcx.struct_tail(pointee);
                 match tail.sty {
                     ty::TyParam(_) | ty::TyProjection(_) => {
-                        assert!(tail.has_param_types() || tail.has_self_ty());
+                        debug_assert!(tail.has_param_types() || tail.has_self_ty());
```
Comment (eddyb) [inline src/librustc/ty/layout.rs]: Checking flags on `Ty` is `O(1)`, why were these changed?

File: src/librustc/ty/sty.rs [hunk 2]
```diff
@@ -708,7 +708,7 @@ impl<'a, 'gcx, 'tcx> ExistentialTraitRef<'tcx> {
     pub fn with_self_ty(&self, tcx: TyCtxt<'a, 'gcx, 'tcx>, self_ty: Ty<'tcx>)
         -> ty::TraitRef<'tcx>  {
         // otherwise the escaping regions would be captured by the binder
-        assert!(!self_ty.has_escaping_regions());
+        debug_assert!(!self_ty.has_escaping_regions());
```
Comment (eddyb) [inline src/librustc/ty/sty.rs]: Same here.

File: src/librustc/ty/sty.rs [hunk 3]
```diff
@@ -1247,7 +1247,7 @@ impl<'a, 'tcx, 'gcx> ExistentialProjection<'tcx> {
                         -> ty::ProjectionPredicate<'tcx>
     {
         // otherwise the escaping regions would be captured by the binders
-        assert!(!self_ty.has_escaping_regions());
+        debug_assert!(!self_ty.has_escaping_regions());
```
Comment (eddyb) [inline src/librustc/ty/sty.rs]: Same here.

**Comment by eddyb** [comment]
This PR seems like it's working around `TraitRef` not having its own precomputed flags?
Maybe we should be interning `TraitRef`s, similar to `Ty`?

Unlike what @varkor said, I doubt many of these are correct, if any, due to how the situations they're checking for (e.g. "no inference variables") can have side-effects without causing an ICE.
E.g. inference variables can get resolved to something that the code assumed to not be there because it 1. looked for the thing explicitly 2. it asserted there are no inference variables.
